---
title: "Test"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
knitr::opts_knit$set(global.par = TRUE)

```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
library(plyr)
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(gplots)
library(gtools)
library(stats)
library(lattice)
library(latticeExtra)
library(grid)
library(gridExtra)
library(Rtsne)
library(stringr)
```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}

## set baseDir if Knitting with RStudio button
baseDir <- paste0(dirname(dirname(getwd())), "/")

## baseDir <- "/Users/sadams/Desktop/Single-Cell-qPCR-Pipeline/single_cell_UCMAIT-master/"

## sets path if run from runAnalysis.R
#currentFile <- rstudioapi::getActiveDocumentContext()$path
#baseDir <- gsub("single_cell_insulin/.*", "single_cell_insulin/", currentFile)

#### source function scripts ####
funcDir <- paste(baseDir, "src/functions/", sep="")
funcFiles <- c("formatRaw.R", 
               "dataLoad.R", 
               "cleanCt.R", 
               "geneSetEnrichment.R", 
               "violinPlot.R", 
               "heatmap2_custom.R",
               "heatmap.3.R",
               #"clusterFilter.R",
               "clusterFilter_1st.R",
               "clusterFilter_2nd.R",
               "tableBarPlots.R",
               "clusterReport.R",
               "reportViolins.R",
               "reportTSNE.R",
               "plotTSNE.R")
funcFiles <- paste(funcDir, funcFiles, sep="")
for(file in funcFiles){
  source(file)
}

```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
## load data
## The options for variables are: Project = "UCM", Species = "HU" or "MS", FreshorFrozen = "FR", "FZ" or "All", Panel = "1", "3", or "1 and 3", "CellID".
## Troubleshoot = "TRUE" or "FALSE"; a "TRUE" here means, it will print all of the "print" statements to make it easier to track issues with the code.
## CombineCommonGenes = "TRUE" here means the code will find the common genes in the panels, it will then average out the two values for each cell, then use the average value and drop both of the single measurements.If you'd prefer to see the two measurements for each cell, then use the "FALSE" setting. This will put an underscore and a number next to the duplicate genes in the final PDF.
## The options should be spelled exactly as listed above, and are always characters (meaning it must be put in quotes) unless it's TRUE/FALSE, otherwise the grep() functions in dataLoad.R might fail.

ctTable <- dataLoad(Project = "T1D", Species = "MS", FreshorFrozen = "All", Panel = "1 and CellID", Troubleshoot = TRUE, CombineCommonGenes = TRUE, RemoveDonor = c(), RemoveSample = c())

## calculate log2Ex, reformat ctTable, and normalize values
ctNorm <- cleanCt(ctTable, summaryOutput=T, cumExpCutoff = F, normGene = "none")
ctNorm[is.na(ctNorm)] <- 0

```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
## cluster and filter
ctClust <- clusterFilter_1st(ctNorm, testK = T, numCenters = 6, plotHeatmap=T, plotClustOnly=F,
                          heatmapFactor = "kmeans.cluster", heatmapColorBy = c(),
                          heatmapTissueLabel = "NOD Doublets",
                          fisherTests = c(),
                         # fisherTests = NULL,
                          cumulativeExpHist = T, filterClusters = F, clustersToRemove = c())
```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
## cluster again to check performance
###SMS edit: If you want to use "filterClusters" to remove a cluster you still have to specify which cluster to remove in "clusterToRemove".
ctClust <- clusterFilter_2nd(ctClust, testK = T, numCenters = 8, plotHeatmap=T, plotClustOnly=F,
                          heatmapFactor = "kmeans.cluster",
                          heatmapColorBy = c(),
                          heatmapTissueLabel = "filtered T1D Initial Draw",
                          fisherTests = c(),
                         # fisherTests = NULL,
                          cumulativeExpHist = T, filterClusters = F, clustersToRemove = c())

```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=TRUE, comment=NA}
#### t-sne reports ####
###This function has been updated from LG's original. The colorby vector can take the following options: "kmeans.cluster", "probe", "patient", "age", "FRorFZ", "tissue", "IBD_Status", "Gene_List". "kmeans.cluster" = tSNE colored by the kmeans clustering numbers, "probe" refers to the Patient IDs. "patient" refers to the cohort the donor is ###from,"tissue source" refers to the tissue the MAIT cell was isolated from (inflamed, uninflamed or blood). IBD_Status is whether the donor was a UC or Crohn's patient. ###"Gene_List" is a bit complicated. It is an option if you want to see the tSNE colored by expression level of specific genes. If you would like to see the expression level on ###the tSNE, you need to include "Gene_List" in the "color_by" vector and the list of genes you'd like to see in the "Genes" input.

ctClust <- plotTSNE(ctClust, colorby = c("kmeans.cluster", "Gene_List"), Genes = c("ACTA2","ACVR1","ADGRE1","ANGPT1","ANPEP","AIM2","BCL2","BCL6","BMP5","BMP7",
"CCL13","CCR1","CCR2","CCR3","CCR4","CCR5","CCR6","CCR7","CD14","CD24A","CD28",
"CD36","CD3E","CD4","CD40","CD44","CD74","CD80","CD83","CD86","CD8A","CEACAM1",
"CLEC7A","COL11A1","COL1A1","COL1A2","CSF1","CSF1R","CSF2RA","CSF2RB","CXCL10",
"CXCL13","CXCR3","CXCR4","DES","EGFR","FAP","FCGR1","FGFR1","FGFR3","FGR","FYN",
"GAPDH","GATA4","GCG","GFAP","GHRL","GM13889","GSK3A","GSK3B","H2-AA","H2-DMA",
"HIF1A","HPRT","IAPP","ICAM1","ICAM2","ICOS","ICOSL","IFIT1","IFIT3","IFI44",
"IFI44L","IFNG","IFNGR1","IGF1","IGF2","IL-21","IL1A","IL1B","IL1R2","IL2","IL2RA",
"IL3","IL4","IL4RA","IL5","IL5RA","IL6","IL7","IL7R","IL10","IL12B","IL12RB","IL17A",
"IL18R1","IL25","IL27","IL27R","IL34","INS1","INS2","IRF1","IRF2","IRF4","IRF7",
"ISG15","ITGAX","ITGB1","JAK1","JAK2","KDR","KLF5","LCK","LEPR","LY6E","LY75",
"MAP2K6","MAPK8","MMP1A","MMP2","MMP3","MMP9","MX1","NFATC1","NFKB1","NLRP3",
"NUR77","OAS1B","OAS2","OASL1","PD1","PDL-1","PDGFA","PDGFB","PDGFRB","PDPN",
"PECAM1","PPARA","PPARG","PPARGC1A","PPY","PTEN","PTGS2","PTK2","RSAD2","RSPO1",
"SELE","SFRP1","SOCS3","SPP1","SST","STAT1","STAT3","STAT4","STAT5","TBX21",
"TEK","TGFB1","TGFBR2","TIMP1","TIMP2","TLR3","TLR4","TLR7","TLR9","TNC",
"TNF","TNFAIP3","TNFRSF1A","TNFRSF1B","TNFSF11","TRAF2","VAV1","VCAM1","VEGFA",
"VEGFB","WNT2B","WNT4","ZAP70","ZEB2"))
```

```{r fig.width=22, fig.height=25, warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, comment=NA}
#### violins ####
## Differentially expressed genes between cluster
violinPlot(ctGenes=ctClust, 
           byFactor="kmeans.cluster", 
           factorOrder=c(1:8),
           groupLabel="clusters", 
           extraLabel="",
           dotSize = 1.5, 
           dotAlpha = 0.3)

##  Differentially expressed genes between patients
#violinPlot(ctGenes=ctClust,
#           byFactor="patient",
#           factorOrder= NULL,
#           groupLabel="patients",
#           extraLabel="",
#           dotSize = 1.5,
#           dotAlpha = 0.3)

##  Differentially expressed genes between probes
#violinPlot(ctGenes=ctClust,
#           byFactor   = "probe",
#           factorOrder = NULL,
#           groupLabel = "probe",
#           extraLabel = "",
#           dotSize    = 1.5,
#           dotAlpha   = 0.3)


# ##  Differentially expressed genes between tissues
# violinPlot(ctGenes=ctClust,
#            byFactor="age",
#            factorOrder=c("infl", "uninfl", "blood"),
#            groupLabel="tissues",
#            extraLabel="for human samples",
#            dotSize = 1.5,
#            dotAlpha = 0.3)

##  Differentially expressed genes between phenotypes
# violinPlot(ctGenes=ctClust,
#            byFactor="cellSource",
#            factorOrder=c("UCM","UCM-PED", "NBD"), #removed "NBD" from here
#            groupLabel="phenotypes",
#            extraLabel="for human samples",
#            dotSize = 1.5,
#            dotAlpha = 0.3)

#### bottom ####
```
